src/main/scala/com/fulcrumgenomics/bam/Pileup.scala

```diff
@@ -180,30 +168,44 @@ class PileupBuilder(dict: SAMSequenceDictionary,
     val buffer = new ListBuffer[PileupEntry]()
     iterator.bufferBetter
       .dropWhile(r => r.getReferenceIndex < refIndex)
-      .takeWhile(r => r.getReferenceIndex == refIndex && r.getAlignmentStart <= pos)
+      .takeWhile(r => r.getReferenceIndex == refIndex && r.getAlignmentStart <= pos + 1)
       .filterNot(r => r.getReadUnmappedFlag)
+      .filter(r => r.getAlignmentStart <= pos || startsWithInsertion(r))
       .filter(r => r.getAlignmentEnd >= pos)
       .filter(r => this.readFilters.forall(f => f(r)))
       .foreach { rec =>
-        val entry = rec.getReadPositionAtReferencePosition(pos) match {
-          case 0 =>
-            // Must be deleted in the read
-            val delPos = rec.getReadPositionAtReferencePosition(pos, true)
-            val entry = DeletionEntry(rec, delPos-1)
-            if (accepts(entry)) buffer += entry
-          case p =>
-            val entry = BaseEntry(rec, p-1)
-            if (accepts(entry)) buffer += entry
-
-            // Also check to see if the subsequent base represents an insertion
-            if (p < rec.getReadLength - 1 && rec.getReferencePositionAtReadPosition(p+1) == 0) {
-              val entry = InsertionEntry(rec, p+1)
+        if (rec.getAlignmentStart == pos + 1) {
+          // Must be an insertion at the start of the read
+          val entry = new InsertionEntry(rec, 0)
+          if (accepts(entry)) buffer += entry
+        }
+        else {
+          rec.getReadPositionAtReferencePosition(pos) match {
+            case 0 =>
+              // Must be deleted in the read
+              val delPos = rec.getReadPositionAtReferencePosition(pos, true)
+              val entry = DeletionEntry(rec, delPos-1)
               if (accepts(entry)) buffer += entry
-            }
+            case p =>
+              val entry = BaseEntry(rec, p-1)
+              if (accepts(entry)) buffer += entry
+
+              // Also check to see if the subsequent base represents an insertion
+              if (p < rec.getReadLength - 1 && rec.getReferencePositionAtReadPosition(p+1) == 0) {
+                val entry = InsertionEntry(rec, p)
+                if (accepts(entry)) buffer += entry
+              }
+          }
         }
       }
 
     Pileup(refName=refName, refIndex=refIndex, pos=pos, pile=buffer)
   }
+
+  /** Returns true if the read is mapped and the first non-clipping operator is an insertion. */
+  private def startsWithInsertion(r: SAMRecord) = {
+    !r.getReadUnmappedFlag &&
+      r.getCigar.getCigarElements.dropWhile(_.getOperator.isClipping).next().getOperator == CigarOperator.INSERTION
```

Comment:
Is it safer to replace `.next().getOperator == CigarOperator.INSERTION` with `.exists(_.getOperator == CigarOperator.INSERTION)` even though it should have a non-clipping operator?